### PR TITLE
FIX: Release CPU/GPU memory at the end of the Program 

### LINKED ISSUE #3536: CPU free memory error
Allocation:
```
92: I0816 20:35:48.884296 3000099776 buddy_allocator.cc:238] Split block (0x10963d000, 536870912) into
92: I0816 20:35:48.884300 3000099776 buddy_allocator.cc:242] Left block (0x10963d000, 4096)
92: I0816 20:35:48.884305 3000099776 buddy_allocator.cc:249] Insert right block (0x10963e000, 536866816)
92: I0816 20:35:48.884318 3000099776 buddy_allocator.cc:54] Allocate 8 bytes from chunk size 4096
92: I0816 20:35:48.884322 3000099776 buddy_allocator.cc:73] Allocation from existing memory block 0x10963e000 at address 0x10963e040
92: I0816 20:35:48.884325 3000099776 buddy_allocator.cc:238] Split block (0x10963e000, 536866816) into
92: I0816 20:35:48.884328 3000099776 buddy_allocator.cc:242] Left block (0x10963e000, 4096)
92: I0816 20:35:48.884331 3000099776 buddy_allocator.cc:249] Insert right block (0x10963f000, 536862720)
92: I0816 20:35:48.884338 3000099776 buddy_allocator.cc:54] Allocate 32 bytes from chunk size 4096
92: I0816 20:35:48.884342 3000099776 buddy_allocator.cc:73] Allocation from existing memory block 0x10963f000 at address 0x10963f040
92: I0816 20:35:48.884346 3000099776 buddy_allocator.cc:238] Split block (0x10963f000, 536862720) into
92: I0816 20:35:48.884349 3000099776 buddy_allocator.cc:242] Left block (0x10963f000, 4096)
92: I0816 20:35:48.884352 3000099776 buddy_allocator.cc:249] Insert right block (0x109640000, 536858624)
```

Free:

```c++
92: I0816 20:35:48.884398 3000099776 buddy_allocator.cc:30] BuddyAllocator Disconstructor makes sure that all of thesehave actually been freed
92: I0816 20:35:48.884402 3000099776 buddy_allocator.cc:34] Free from block (0x109640000, 536870912)
```

There is a memory bug since free from `0x109640000` and size is `536870912`, it should not be`536858624`

### LINKED ISSUE #3537: tensor didn't free memory at the end of the program
@zchen0211  In scatter_test and gather_test, tensor didn't free memory at the end of the program

### PR DESCRIPTION
**Pull request by gangliao**
After unit test is passed, release memory!

```bash
[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (600 ms total)
[  PASSED  ] 4 tests.
I0816 19:44:47.157642 3000099776 buddy_allocator.cc:30] BuddyAllocator Disconstructor makes sure that all of these have actually been freed
I0816 19:44:47.157649 3000099776 buddy_allocator.cc:34] Free from block (0x700960000, 1818944052)
I0816 19:44:47.159529 3000099776 buddy_allocator.cc:30] BuddyAllocator Disconstructor makes sure that all of these have actually been freed
I0816 19:44:47.159555 3000099776 buddy_allocator.cc:34] Free from block (0x11419c000, 536870912)
```

fix #3536 
fix #3537

### CODE REVIEW

File: paddle/operators/gather_test.cc [hunk 1]
```diff
@@ -45,4 +45,8 @@ TEST(Gather, GatherData) {
 
   for (int i = 0; i < 4; ++i) EXPECT_EQ(p_output[i], i + 4);
   for (int i = 4; i < 8; ++i) EXPECT_EQ(p_output[i], i - 4);
+
```
Comment (gangliao) [inline paddle/operators/gather_test.cc]: @zchen0211 return tensor memory back to buddy allocator for free.

File: paddle/memory/memory.cc [hunk 2]
```diff
@@ -13,22 +13,33 @@ See the License for the specific language governing permissions and
...
-  }
-  return a;
+using BuddyAllocator = detail::BuddyAllocator;
+
+std::once_flag cpu_allocator_flag;
+std::once_flag gpu_allocator_flag;
+
+BuddyAllocator* GetCPUBuddyAllocator() {
+  static std::unique_ptr<BuddyAllocator, void (*)(BuddyAllocator*)> a{
+      nullptr, [](BuddyAllocator* p) { delete p; }};
```
Comment (QiJune) [inline paddle/memory/memory.cc]: Why here need to pass a lambda function, unique_ptr will delete the pointer automaticlly

Comment (gangliao) [inline paddle/memory/memory.cc]: You are right! Deleter could be default one.

Comment (gangliao) [inline paddle/memory/memory.cc]: Done

**Pull request merged**